ambry-replication/src/main/java/com/github/ambry/replication/BackupCheckerThread.java

```diff
@@ -0,0 +1,165 @@
+    super(threadName, findTokenHelper, clusterMap, correlationIdGenerator, dataNodeId, connectionPool,
+        replicationConfig, replicationMetrics, notification, storeKeyConverter, transformer, metricRegistry,
+        replicatingOverSsl, datacenterName, responseHandler, time, replicaSyncUpManager, skipPredicate,
+        leaderBasedReplicationAdmin);
+  }
+
+  /**
+   *  This checks for GDPR compliance.
+   *  If a blob has been deleted from on-prem servers, it must be absent in backups.
+   *  If it is in TEMP or PERM state, then backup is behind servers & DELETE has not propagated to the backup.
+   * @param onPremBlob the {@link MessageInfo} that will be transformed into a delete-operation
+   * @param remoteReplicaInfo The remote replica that is being replicated from
+   */
+  @Override
+  protected void applyDelete(MessageInfo onPremBlob, RemoteReplicaInfo remoteReplicaInfo) {
+    EnumSet<MessageInfoType> acceptableBackupBlobStates = EnumSet.noneOf(MessageInfoType.class);
+    EnumSet<StoreErrorCodes> acceptableStoreErrorCodes = EnumSet.of(StoreErrorCodes.ID_Deleted,
+        StoreErrorCodes.ID_Deleted_Permanently, StoreErrorCodes.ID_Not_Found);
+    checkBackup(onPremBlob, remoteReplicaInfo, acceptableBackupBlobStates, acceptableStoreErrorCodes);
+  }
+
+  /**
+   * This method checks if permanent blob is permanent in backups as well.
+   * This method does not expect any errors when retrieving the blob from store.
+   * @param onPremBlob the {@link MessageInfo} that will be transformed into a TTL update
+   * @param remoteReplicaInfo The remote replica that is being replicated from
+   */
+  @Override
+  protected void applyTtlUpdate(MessageInfo onPremBlob, RemoteReplicaInfo remoteReplicaInfo) {
+    EnumSet<MessageInfoType> acceptableBackupBlobStates = EnumSet.of(MessageInfoType.TTL_UPDATE);
+    EnumSet<StoreErrorCodes> acceptableStoreErrorCodes = EnumSet.noneOf(StoreErrorCodes.class);
+    checkBackup(onPremBlob, remoteReplicaInfo, acceptableBackupBlobStates, acceptableStoreErrorCodes);
+  }
+
+  /**
+   * This method checks if an un-deleted blob has been un-deleted from backups as well.
+   * The backup could be lagging behind servers and the blob could be temp or perm state.
+   * This is acceptable as well since un-delete takes the blob back to one of these states.
+   * And if we recover from a lagging backup in case of a disaster, we are consistent from a user's point of view.
+   * TODO: If we the blob was permanent before un-delete, then it should be permanent after un-delete.
+   * TODO: Same for a temporary blob. Distinguish between these two cases.
+   * @param onPremBlob the {@link MessageInfo} that will be transformed into an undelete
+   * @param remoteReplicaInfo The remote replica that is being replicated from
+   */
+  @Override
+  protected void applyUndelete(MessageInfo onPremBlob, RemoteReplicaInfo remoteReplicaInfo) {
+    EnumSet<MessageInfoType> acceptableBackupBlobStates = EnumSet.of(MessageInfoType.PUT,
```

Comment:
True, for # 2 it means that backup is behind replica but its not data loss.

Reply:
#2, we only can know the backup doesn't have the undelete record.
We are not sure it's because of replica lag or there is bug that we don't support undelete yet.

Reply:
yeah, fixed